arviz/stats/stats_utils.py

```diff
@@ -0,0 +1,322 @@
+        """General ufunc for multi-output function."""
+        element_shape = ary.shape[:-n_dims]
+        if out is None:
+            out = tuple(np.empty(element_shape) for _ in range(n_output))
+        else:
+            raise_error = False
+            if isinstance(out, tuple):
+                out_shape = tuple(item.shape for item in out)
+                correct_shape = tuple(element_shape for _ in range(n_output))
+                if out_shape != correct_shape:
+                    raise_error = True
+            else:
+                raise_error = True
+            if raise_error:
+                msg = "Shapes incorrect for `out`: {}.".format(out_shape)
+                msg += " Correct shapes are {}".format(correct_shape)
+                raise TypeError(msg)
+        for idx in np.ndindex(element_shape):
+            ary_idx = ary[idx].ravel() if ravel else ary[idx]
+            results = func(ary_idx, *args, **kwargs)
+            for i, res in enumerate(results):
+                out[i][idx] = np.asarray(res)[index]
+        return out
+
+    if n_output > 1:
+        ufunc = _multi_ufunc
+    else:
+        ufunc = _ufunc
+
+    update_docstring(ufunc, func, n_output)
+    return ufunc
+
+
+def wrap_xarray_ufunc(
+    ufunc, dataset, *, ufunc_kwargs=None, func_args=None, func_kwargs=None, **kwargs
+):
+    """Wrap make_ufunc with xarray.apply_ufunc.
+
+    Parameters
+    ----------
+    ufunc : callable
+    dataset : xarray.dataset
+    ufunc_kwargs : dict
+        Keyword arguments passed to `make_ufunc`.
+            - 'n_dims', int, by default 2
+            - 'n_output', int, by default 1
+            - 'index', slice, by default Ellipsis
+            - 'ravel', bool, by default True
+    func_args : tuple
+        Arguments passed to 'ufunc'.
+    func_kwargs : dict
+        Keyword arguments passed to 'ufunc'.
+    **kwargs
+        Passed to xarray.apply_ufunc.
+    Return
+    ------
+    xarray.dataset
+    """
+    if ufunc_kwargs is None:
+        ufunc_kwargs = {}
+    if func_args is None:
+        func_args = tuple()
+    if func_kwargs is None:
+        func_kwargs = {}
+
+    n_output = ufunc_kwargs.pop("n_output", 1)
+    callable_ufunc = make_ufunc(ufunc, n_output=n_output, **ufunc_kwargs)
+
+    if "input_core_dims" in kwargs:
+        input_core_dims = kwargs.pop("input_core_dims")
+    else:
+        input_core_dims = tuple(("chain", "draw") for _ in range(len(func_args) + 1))
+    if "output_core_dims" in kwargs:
+        output_core_dims = kwargs.pop("output_core_dims")
+    else:
+        output_core_dims = tuple([] for _ in range(n_output))
+
+    return apply_ufunc(
+        callable_ufunc,
+        dataset,
+        *func_args,
+        kwargs=func_kwargs,
+        input_core_dims=input_core_dims,
+        output_core_dims=output_core_dims,
+        **kwargs
+    )
+
+
+def update_docstring(ufunc, func, n_output=1):
+    """Update ArviZ generated ufunc docstring."""
+    module = ""

```

Comment:
Do you think this would be easier to write with fstrings?

Reply:
Can we use fstrings with py3.5?

Reply:
no, we need Python > 3.6

Reply:
I'm still curious though, would it be easier with fstrings? If is it I'd like to make an issue that if we drop 3.5 support we can refactor this

Reply:
I don't think so. This is quite explicit now. Even with fstrings it woukd need to be explicit.

And this is quite minor thing, only good after people start to use this more (ufunc wrapper).